plugins/repository-s3/src/test/java/org/opensearch/repositories/s3/async/AsyncUploadUtilsTests.java

```diff
@@ -0,0 +1,238 @@
+    public void testMultipartUpload() {
+        CompletableFuture<CreateMultipartUploadResponse> createMultipartUploadRequestCompletableFuture = new CompletableFuture<>();
+        createMultipartUploadRequestCompletableFuture.complete(CreateMultipartUploadResponse.builder().uploadId("uploadId").build());
+        when(s3AsyncClient.createMultipartUpload(any(CreateMultipartUploadRequest.class))).thenReturn(
+            createMultipartUploadRequestCompletableFuture
+        );
+
+        CompletableFuture<UploadPartResponse> uploadPartResponseCompletableFuture = new CompletableFuture<>();
+        uploadPartResponseCompletableFuture.complete(UploadPartResponse.builder().checksumCRC32("pzjqHA==").build());
+        when(s3AsyncClient.uploadPart(any(UploadPartRequest.class), any(AsyncRequestBody.class))).thenReturn(
+            uploadPartResponseCompletableFuture
+        );
+
+        CompletableFuture<CompleteMultipartUploadResponse> completeMultipartUploadResponseCompletableFuture = new CompletableFuture<>();
+        completeMultipartUploadResponseCompletableFuture.complete(CompleteMultipartUploadResponse.builder().build());
+        when(s3AsyncClient.completeMultipartUpload(any(CompleteMultipartUploadRequest.class))).thenReturn(
+            completeMultipartUploadResponseCompletableFuture
+        );
+
+        CompletableFuture<AbortMultipartUploadResponse> abortMultipartUploadResponseCompletableFuture = new CompletableFuture<>();
+        abortMultipartUploadResponseCompletableFuture.complete(AbortMultipartUploadResponse.builder().build());
+        when(s3AsyncClient.abortMultipartUpload(any(AbortMultipartUploadRequest.class))).thenReturn(
+            abortMultipartUploadResponseCompletableFuture
+        );
+
+        CompletableFuture<Void> resultFuture = asyncUploadUtils.uploadObject(
+            s3AsyncClient,
+            new UploadRequest("bucket", "key", ByteSizeUnit.MB.toBytes(5), WritePriority.HIGH, uploadSuccess -> {
+                // do nothing
+            }, true, 3376132981L),
+            new StreamContext(
+                (partIdx, partSize, position) -> new InputStreamContainer(new ZeroInputStream(partSize), partSize),
+                ByteSizeUnit.MB.toBytes(1),
+                ByteSizeUnit.MB.toBytes(1),
+                5
+            )
+        );
+
+        try {
+            resultFuture.get();
+        } catch (ExecutionException | InterruptedException e) {
+            fail("did not expect resultFuture to fail");
+        }
+
+        verify(s3AsyncClient, times(1)).createMultipartUpload(any(CreateMultipartUploadRequest.class));
+        verify(s3AsyncClient, times(5)).uploadPart(any(UploadPartRequest.class), any(AsyncRequestBody.class));
+        verify(s3AsyncClient, times(1)).completeMultipartUpload(any(CompleteMultipartUploadRequest.class));
+        verify(s3AsyncClient, times(0)).abortMultipartUpload(any(AbortMultipartUploadRequest.class));
+    }
+
+    public void testMultipartUploadCorruption() {
```

Comment:
Lets add more UTs for following cases :

1. Clean up happening after delete 
2. One part of the upload failing , thus failing the overall upload 
3. complete multipart upload failure .

Reply:
We can add this in the follow up tasks. @raghuvanshraj

Reply:
For this [comment](https://github.com/opensearch-project/OpenSearch/pull/7217#discussion_r1236310903). It requires a very long explanation. In short, it is not just event loop threads or future completion threads. Main thread and stream processing threads are also involved. It depends on at what point s3 is invoking asyncIO for transfer, how we have chained futures and how handling of callbacks is being done. First future is executed in main thread, callbacks, subsequent calls and end response in future completion threadpool and event loop threads are mostly managing IO.

Reply:
For this [comment](https://github.com/opensearch-project/OpenSearch/pull/7217#discussion_r1236310903). It requires a very long explanation. In short, it is not just event loop threads or future completion threads. Main thread and stream processing threads are also involved. It depends on at what point s3 is invoking asyncIO for transfer, how we have chained futures and how handling of callbacks is being done. First future is executed in main thread, callbacks, subsequent calls and end response in future completion threadpool and event loop threads are mostly managing IO ops.